transaction/std/src/memo/destination.rs

```diff
@@ -0,0 +1,148 @@
+//! Object for Destination memo type
+
+use core::convert::TryInto;
+use mc_account_keys::AddressHash;
+use mc_transaction_core::MemoPayload;
+
+/// A memo that the sender writes to themself to record details of the
+/// transaction, and attaches to the change TxOut so that they can recover it
+/// later.
+///
+/// This memo is not authenticated -- authentication is performed by confirming
+/// that the TxOut belongs to the change subaddress, which should not be
+/// disclosed to third parties.
+///
+/// The sender of the transaction is able to put fake data in the
+/// DestinationMemo, nothing in the system forces it to match what actually
+/// happened in the transaction. These are records created for Bob to help
+/// himself remember what he did, if he wants to.
+///
+/// The layout of the memo data in 32 bytes is:
+/// [0-16]: sender_address_hash
+/// [16]: num_recipients
+/// [17-24]: fee
+/// [24-32]: total outlay
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct DestinationMemo {
+    /// The address hash of the recipient to whom the payment is attributed
+    address_hash: AddressHash,
+    /// The number of recipients of the transaction (ignoring the change output
+    /// and fee). For a typical transaction, this is one, and the address
+    /// hash refers to that recipient. When there is more than one
+    /// recipient, one of them can be chosen arbitrarily.
+    num_recipients: u8,
+    /// The total fee paid in the transaction (in picomob)
+    ///
+    /// Note: We assume that the high order byte of fee is zero, and use this
+    /// to compress the memo into 32 bytes. For this assumption not to be
+    /// correct, there would have to be a transaction that spends more than
+    /// 1% of all of MOB as the fee, which is considered not an important
+    /// scenario.
+    fee: u64,
+    /// The sum of all outlays of the transaction (in picomob)
+    /// Here outlay means, sum of amounts of outputs that are not change and
+    /// are not the transaction fee.
+    /// For a typical transaction with one recipient, this refers to the amount
+    /// of a single TxOut.
+    ///
+    /// This memo is attached to the change TxOut of the transaction, and from
+    /// the amount of that, and this memo, the owner can determine the sum of
+    /// the inputs spent in the transaction by adding change + fee +
+    /// total_outlay
+    total_outlay: u64,
```

Comment:
Should probably include a note about why it is okay to use a u64 even though the total amount could actually exceed u64